### PR TITLE
Parent Windows get chance to cancel closing before any children close.

### PR DESCRIPTION
**Pull request by danwalmsley**
## What does the pull request do?

Allows apps that use child windows to be able to easily prevent the app from closing.

## What is the current behavior?

If you have child windows open, each child window gets asked if it wants to cancel closing, and if it doesnt, it immediately closes. Child windows are asked before the main window.

## What is the updated/expected behavior with this PR?

Each child window is enumerated with closing event giving the chance to cancel. then the mainwindow is asked if it wants to cancel via closing event.

Only after this has happened are any windows closed or not as the case may be.

This matches the order that WPF does things (the only difference that we enumerate child windows and wpf doesnt)

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

File: src/Avalonia.Controls/Window.cs [hunk 1]
```diff
@@ -515,20 +525,13 @@ private void CloseInternal()
...
             bool canClose = true;
 
             foreach (var (child, _) in _children.ToList())
             {
-                if (!child.HandleClosing())
-                {
-                    child.CloseInternal();
-                }
-                else
+                if (child.ShouldCancelClose())
```
Comment (MarchingCube) [inline src/Avalonia.Controls/Window.cs:534]: Do we actually want to check all windows even if one already cancelled?

Comment (danwalmsley) [reply src/Avalonia.Controls/Window.cs]: I think so, so that you get the closing event?

Comment (MarchingCube) [reply src/Avalonia.Controls/Window.cs]: I guess that with your change to pass args in here it makes sense. So you can tell if it will actually close.

**Review by MarchingCube — APPROVED** [review]
LGTM. Tested locally on win32 to verify as well. Linux was not manually tested (since my VM died somehow).

**Pull request merged**